opc/resource_database_service_instance.go

```diff
@@ -0,0 +1,594 @@
+package opc
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"time"
+
+	opcClient "github.com/hashicorp/go-oracle-terraform/client"
+	"github.com/hashicorp/go-oracle-terraform/database"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceOPCDatabaseServiceInstance() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceOPCDatabaseServiceInstanceCreate,
+		Read:   resourceOPCDatabaseServiceInstanceRead,
+		Delete: resourceOPCDatabaseServiceInstanceDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"edition": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(database.ServiceInstanceStandardEdition),
+					string(database.ServiceInstanceEnterpriseEdition),
+					string(database.ServiceInstanceEnterpriseEditionHighPerformance),
+					string(database.ServiceInstanceEnterpriseEditionExtremePerformance),
+				}, true),
+			},
+			"level": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(database.ServiceInstanceLevelPAAS),
+					string(database.ServiceInstanceLevelBasic),
+				}, true),
+			},
+			"shape": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(database.ServiceInstanceShapeOC3),
+					string(database.ServiceInstanceShapeOC4),
+					string(database.ServiceInstanceShapeOC5),
+					string(database.ServiceInstanceShapeOC6),
+					string(database.ServiceInstanceShapeOC1M),
+					string(database.ServiceInstanceShapeOC2M),
+					string(database.ServiceInstanceShapeOC3M),
+					string(database.ServiceInstanceShapeOC4M),
```

Comment:
it depends how often new Shape's are added - but whitelisting Machine SKU's can be painful if they're added frequently. In some Providers this means we don't validate the Machine SKU and let the remote cloud return an error if it's invalid (or check a regex instead)